pkg/trace/api/profiles.go

```diff
@@ -33,21 +41,73 @@ func profilingEndpoint() string {
 	return profilingURLDefault
 }
 
-// profileProxyHandler returns a new HTTP handler which will proxy requests to the profiling intake.
-// If the URL can not be computed because of a malformed 'site' config, the returned handler will always
+// additionalProfilingEndpoints returns a map of endpoint URLs to a slice of api
+// keys to be used for each endpoint. There is no de-duplication between api
+// keys or between these additional endpoints and the main endpoint.
+func additionalProfilingEndpoints() map[string][]string {
+	if config.Datadog.IsSet(profilingAdditionalEndpointsConfigKey) {
+		return config.Datadog.GetStringMapStringSlice(profilingAdditionalEndpointsConfigKey)
+	}
+	return make(map[string][]string)
+}
+
+// profileProxyHandler returns a new HTTP handler which will proxy requests to the profiling intakes.
+// If the main intake URL can not be computed because of config, the returned handler will always
 // return http.StatusInternalServerError along with a clarification.
 func (r *HTTPReceiver) profileProxyHandler() http.Handler {
-	target := profilingEndpoint()
-	u, err := url.Parse(target)
+	tags := fmt.Sprintf("host:%s,default_env:%s", r.conf.Hostname, r.conf.DefaultEnv)
+	mainEndpoint := mainProfilingEndpoint()
+	u, err := url.Parse(mainEndpoint)
 	if err != nil {
-		log.Errorf("Profile forwarder is OFF because of invalid intake URL: %v", err)
+		log.Errorf("Error parsing main intake URL %s: %v", mainEndpoint, err)
+		return errorHandler(mainEndpoint)
+	}
+	mainProxy := newProfileProxy(r.conf.NewHTTPTransport(), u, r.conf.APIKey(), tags)
+	if mainProxy == nil {
+		log.Errorf("Failed to create reverse proxy for main endpoint %s", mainEndpoint)
+		return errorHandler(mainEndpoint)
+	}
+	proxies := []*httputil.ReverseProxy{mainProxy}
+
+	additionalEndpoints := additionalProfilingEndpoints()
+	for endpoint, apiKeys := range additionalEndpoints {
+		u, err := url.Parse(endpoint)
+		if err != nil {
+			log.Errorf("Error parsing additional intake URL %s: %v", endpoint, err)
+			continue
+		}
+
+		for _, apiKey := range apiKeys {
+			proxies = append(proxies, newProfileProxy(r.conf.NewHTTPTransport(), u, apiKey, tags))
+		}
+	}
+
+	switch len(proxies) {
+	case 0:
+		return errorHandler(mainEndpoint)
+	case 1:
+		return proxies[0]
+	default:
 		return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
-			msg := fmt.Sprintf("Agent is misconfigured with an invalid intake URL: %q", target)
-			http.Error(w, msg, http.StatusInternalServerError)
+			bodyBytes, err := ioutil.ReadAll(req.Body)
+			if err != nil {
+				msg := "Failed to read request body"
+				http.Error(w, msg, http.StatusInternalServerError)
+				return
+			}
+			outreq := req.Clone(req.Context())
+			outreq.Body = ioutil.NopCloser(bytes.NewReader(bodyBytes))
```

Comment:
Only thing I can think of improving here is to create a `copyingReader` where first read works in a streaming manner (but saving read bytes in a byte buffer). Then consecutive reads can `rewind` and read from the byte buffer.

The big advantage with this is that if someone happens to send a 2GB profile, the intake can answer with `413 Request Entity Too Large` almost immediately and we never actually load 2GB of things into memory.

Example usage:

```
bodyReader = copyingReader(req.Header.ContentLength, req.Body)
outreq.Body = bodyReader
(...)
outreq.Body = bodyReader.rewind()
```

Reply:
This is a neat idea, but it assumes that we always get a `Content-Length` header and we can trust the client to set it correctly. Is it possible for any of the clients to use `Transfer-Encoding: chunked`?